006-immediate-close.md

```diff
@@ -0,0 +1,200 @@
+is admittedly potential for bugs, however it is easy to spot such bugs as long as one
+is aware of their potential. When one need to access the object after calling a
+callback, the solution is simple (use a `Timer` with zero timeout or a class built
+specifically for this purpose, perhaps with guaranteed LIFO processing).
+
+With these guarantees, the `Printer` class just like `Timer` provides a clear and
+safe API. The design and guarantees propagates upward in the application simplifying
+code all over the place.
+
+Having demonstrated the usefulness of the described design pattern, we come the the
+realization that the `Timer` class cannot be implemented directly with `uv_timer_t`
+because `uv_close` wants the user to keep the `uv_timer_t` memory around until it
+calls the close callback. The `Timer` implementation can only work around this issue,
+by dynamically allocating the `uv_timer_t` object and having the close callback finally
+release the allocation.
+
+## User API
+
+For the API, the following is proposed:
+- The return type of `uv_close` is changed from `void` to `int`.
+- If the callback passed to `uv_close` is not null, the return value is 0 and the
+  behavior is otherwise unchanged. This means that the close callback will be called
+  when the handle memory can be released; callbacks for outstanding any asynchronous
+  requests such as `uv_connect_t` and `uv_write_t` will be called before the close
+  callback with `UV_ECANCELED` status.
+- If the callback passed to `uv_close` is null, the function either closes the
+  handle immediately and returns zero, or if immediate close is not possible does
+  nothing and returns `UV_EBUSY`. In the latter case, the application could still
+  call `uv_close` with non-null callback.
+
+It is believed that the majority of applications pass non-null close callback, and for
+these the change is backward-compatible (disregarding possible warnings about unused
+return value).
+
+Applications which call `uv_close` with null callback will generally break, for example
+with `UV_EBUSY` result the close will never occur at all. The current documentation does
+not specifically mention passing null callback, but it does say that "close_cb will be
+called asynchronously after this call" which implicitly requires it to not be null.
+
+An alternative approach would to return `UV_INPROGRESS` and start closing asynchronously
+when `uv_close` is called with null callback but the close cannot be immediate. This
+change would be more backward compatible with applications currently calling `uv_close`
+with null callback. However, it has the issue that an asynchronous close might start
+without a callback so the application would need the ability to set it after caling
+`uv_close`.
+
+Note that if immediate close is successful, callbacks of outstanding requests are
+*not* called. Since the application has closed the handle it is clearly not interested
```

Comment:
maybe call them with ECANCELD?

Reply:
I have updated the proposal to mention this possibility, but as noted a decision should not be needed right now.